Change PartyAndCertificate to an aggregate class

rnicoll opened the pull request:
Change PartyAndCertificate to an aggregate class instead of a subclass of Party. This
reduces the changes compared to M11, as well as avoiding risk of accidental serialization
of a PartyAndCertificate (which may be very large) where a Party is expected.

Cleaned up initial nodes known to the identity service, in particular mock nodes now know
about themselves; previously full nodes registered themselves but mock nodes did not.

mnesbit commented on core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt:
```diff
@@ -52,24 +50,24 @@ interface IdentityService {
      * used in preference where possible.
      */
-    fun getAllIdentities(): Iterable<Party>
+    fun getAllIdentities(): Iterable<PartyAndCertificate>
```
This seems fair, but I think we need a couple more methods. Specifically:
fun getPublicParty(anonymousParty: AnonymousParty): Party which tracks up the PKI certpath till it hits a public Party in the network map.
fun getPartyAndCertificate(anonymousParty: AnonymousParty): PartyAndCerificate? which tries to retrieve the stored certs for any party, public, anonymous, etc

rnicoll replied:
`partyFromAnonymous` is the first one. I'm adding `certificateFromParty` to match the naming scheme used elsewhere in the interface.

rnicoll commented:
`PartyAndCertificate` is now a data class. I've added documentation of the equality test reasoning (it also seems to break tests if I take this off, hadn't tried before without it, and I should figure out why as that's unexpected).

I've also reworked the documentation on `IdentityService` to describe what it actually does, which is quite different to what it used to say, and to mention what we mean by well known identity.
